src/Symfony/Component/Messenger/Middleware/LockMiddleware.php

```diff
@@ -0,0 +1,68 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Middleware;
+
+use Symfony\Component\Lock\LockFactory;
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Message\LockableMessage;
+use Symfony\Component\Messenger\Stamp\LockStamp;
+use Symfony\Component\Messenger\Stamp\ReceivedStamp;
+
+final class LockMiddleware implements MiddlewareInterface
+{
+    private ?LockFactory $lockFactory = null;
+
+    public function __construct(
+        ?LockFactory $lockFactory,
+    ) {
+        $this->lockFactory = $lockFactory;
+    }
+
+    public function handle(Envelope $envelope, StackInterface $stack): Envelope
+    {
+        if (null === $this->lockFactory) {
+            return $stack->next()->handle($envelope, $stack);
+        }
+
+        $message = $envelope->getMessage();
+
+        // If we're trying to dispatch a lockable message.
+        if ($message instanceof LockableMessage && null === $envelope->last(ReceivedStamp::class)) {
+            $key = $message->getKey();
+
+            if (null !== $key) {
+                // The acquire call must be done before stamping the message
+                // in order to have the full state of the key in the stamp.
+                $canAcquire = $this->lockFactory->createLockFromKey($key, autoRelease: false)->acquire();
+
+                $envelope = $envelope->with(new LockStamp($key));
+                if (!$canAcquire) {
+                    return $envelope;
+                }
+            }
+        }
+
+        try {
+            $envelope = $stack->next()->handle($envelope, $stack);
+        } finally {
```

Comment:
What if message failed and is retried? :thinking:

Reply:
Failure/Retry are handled by the SendFailedMessageForRetryListener
It listen this event https://github.com/symfony/symfony/blob/3fdc2c99d101ad69dd5c99ed03d8d64ad9608631/src/Symfony/Component/Messenger/Worker.php#L199 

But I "handle" the lock before
https://github.com/symfony/symfony/blob/3fdc2c99d101ad69dd5c99ed03d8d64ad9608631/src/Symfony/Component/Messenger/Worker.php#L163

So I don't know at the moment if the message will be retried or not...

I prefer to release the lock rather than riskying keeping it forever because the message is retry. I don't see other solution so far. There was already another chance to have some duplication in the queue, if the first lock reach the end of the TTL. We cannot be sure there is not a big waiting line in the queue some time...

I don't see the LockMiddleware as a way to have absolute 0 duplication in the queue but as a way to reduce the duplication as much as technically possible. But I'm open to any suggestion to improve this.

Reply:
I agree.